spring/src/main/java/org/axonframework/spring/config/SpringComponentRegistry.java

```diff
@@ -0,0 +1,582 @@
+    private final List<Class<? extends ConfigurationEnhancer>> invokedEnhancers = new CopyOnWriteArrayList<>();
+
+    private ConfigurableListableBeanFactory beanFactory;
+
+    /**
+     * Constructs a {@code SpringComponentRegistry} with the given {@code listableBeanFactory}. The
+     * {@code listableBeanFactory} is used to discover all beans of type {@link ConfigurationEnhancer}.
+     *
+     * @param listableBeanFactory The bean factory used to discover all beans of type {@link ConfigurationEnhancer}.
+     */
+    @Internal
+    public SpringComponentRegistry(@Nonnull ListableBeanFactory listableBeanFactory) {
+        Objects.requireNonNull(listableBeanFactory, "The listableBeanFactory may not be null.");
+        this.enhancers.addAll(listableBeanFactory.getBeansOfType(ConfigurationEnhancer.class).values());
+    }
+
+    @Override
+    public <C> ComponentRegistry registerComponent(@Nonnull ComponentDefinition<? extends C> definition) {
+        if (!(definition instanceof ComponentDefinition.ComponentCreator<? extends C> creator)) {
+            // The compiler should avoid this from happening.
+            throw new IllegalArgumentException("Unsupported component definition type: " + definition);
+        }
+
+        // We need to buffer these components, because they may depend on components that aren't
+        // registered yet. We will register these in the application context just-in-time.
+        Component<? extends C> component = creator.createComponent();
+        if (components.contains(component.identifier())) {
+            throw new ComponentOverrideException(creator.type(), creator.name());
+        }
+
+        components.put(component);
+        return this;
+    }
+
+    @Override
+    public <C> ComponentRegistry registerDecorator(@Nonnull DecoratorDefinition<C, ? extends C> definition) {
+        if (!(definition instanceof DecoratorDefinition.CompletedDecoratorDefinition<C, ? extends C> decoratorRegistration)) {
+            // The compiler should avoid this from happening.
+            throw new IllegalArgumentException("Unsupported decorator definition type: " + definition);
+        }
+
+        logger.debug("Registering decorator definition: [{}]", definition);
+        this.decorators.add(decoratorRegistration);
+        return this;
+    }
+
+    @Override
+    public boolean hasComponent(@Nonnull Class<?> type) {
+        // Checks both the local Components as the BeanFactory,
+        //  since the ConfigurationEnhancers act before component registration with the Application Context.
+        return components.contains(new Component.Identifier<>(type, null))
+                || beanFactory.getBeanNamesForType(type).length > 0;
+    }
+
+    @Override
+    public boolean hasComponent(@Nonnull Class<?> type,
+                                @Nullable String name) {
+        Assert.notNull(name, () -> "Spring does not allow the use of null names for components.");
```

Comment:
Wouldn't any component of that type match by our logic?

Reply:
Err, yea, you're right. Let me redirect a `name == null` invocation to the `hasComponent(Class)` method instead. That should cover it sufficiently.